### ISSUE TITLE
Chat templating and tool-call handling belong in our Rust: upstream declined the binding, with reasons

### ISSUE THREAD

**Issue by pixie79** [OffeneDatenmodellierung/Roteiro#492]
`llama-cpp-2` exposes `apply_chat_template`, which wraps `llama_chat_apply_template` — and `llama.h:1197` states plainly:

> `NOTE: This function does not use a jinja parser. It only support a pre-defined list of template.`

`llm_chat_detect_template` (`src/llama-chat.cpp:89`) substring-matches; anything containing `<|im_start|>` becomes `CHATML`, rendered by eight lines of C++. Verified live: `apply_chat_template` given qwen3-32b's own 4,100-byte Jinja template — first line `{%- if tools %}` — returns **153 bytes**. The Jinja never runs.

## The capability is already compiled into our binary

`common/chat.h:325` declares `common_chat_templates_apply`, taking `common_chat_templates_inputs` (`chat.h:250`) with `tools`, `tool_choice`, `parallel_tool_calls`, `use_jinja` — and returning `common_chat_params` with a rendered `prompt`, a **generated `grammar`**, `grammar_triggers` and a `parser`.

`common/chat.cpp` and a first-party Jinja engine (`common/jinja/*.cpp`) are both in libcommon (`common/CMakeLists.txt:66,109-120`), and `rto-llama` already links it — `crates/rto-llama/Cargo.toml:19` does not set `default-features = false`, and llama-cpp-2's defaults include `common`.

There is simply **no binding**:

```
grep -rn "common_chat" llama-cpp-2-0.1.154/src/            → 0
grep -n  "chat"        llama-cpp-sys-2-0.1.154/wrapper_common.h    → 0
grep -n  "common_chat" wrapper_common.cpp                  → 0
```

**Nothing needs vendoring or a version bump.** The code is in the archive; there is no door.

## The ask

A C-ABI shim in `wrapper_common.{h,cpp}` — the same pattern as the existing `llama_rs_json_schema_to_grammar` — exposing `common_chat_templates_init` / `common_chat_templates_apply` / `common_chat_templates_free`, marshalling `tools`, `tool_choice`, `parallel_tool_calls` in, and `prompt` + `grammar` + `grammar_triggers` out as JSON strings.

## Why it is worth raising even though we are not blocked on it

It would hand us a **per-model grammar generated by llama.cpp itself** — strictly better than the envelope grammar #485's PR 2 builds by hand — and it would make **PR 3 unnecessary**, because the dialect problem in #489 (`qwen3-coder-30b-a3b` and `qwen3.8-27b` trained on XML while our parser reads JSON) is exactly what `common_chat`'s per-template parser already solves upstream.

## Do not block on it

File it, note it in the ADR as the intended end state, ship #485's PRs 1–3 regardless. If it lands, PR 3 collapses into it.

Two things to settle before filing upstream: confirm the shim pattern against the current `llama-cpp-rs` main rather than the pinned 0.1.154, and check whether an issue or PR for this already exists there.

**Comment by pixie79** [comment]
## Upstream tickets exist — linking rather than filing a duplicate

`llama-cpp-sys-2` / `llama-cpp-2` is [utilityai/llama-cpp-rs](https://github.com/utilityai/llama-cpp-rs). Two open issues there already cover this, and one of them is the exact finding in this issue, arrived at independently:

| upstream | state | |
|---|---|---|
| [utilityai/llama-cpp-rs#1096](https://github.com/utilityai/llama-cpp-rs/issues/1096) | open, 2026-08-11 | **`apply_chat_template` silently gives the wrong prompt for most models** — quotes the same `llama-chat.cpp` heuristic, and reports running 18 models' templates through it |
| [utilityai/llama-cpp-rs#864](https://github.com/utilityai/llama-cpp-rs/issues/864) | open, 2025-11-19 | Incomplete Rust bindings for llama.cpp tool calling — the broader ask |

#1096 was filed eight days ago with **zero comments**. Our independent measurement is worth adding to it: qwen3-32b's own 4,100-byte Jinja template, first line `{%- if tools %}`, returns **153 bytes** through `apply_chat_template`. The Jinja never runs.

## The history is worse than this issue assumed, and changes the ask

This issue is framed as *"ask upstream to add a binding"*. It has been added, and then lost:

| | | |
|---|---|---|
| [#977](https://github.com/utilityai/llama-cpp-rs/pull/977) | **merged** 2026-04-06 | *"typed tool definitions, chat template application **with tools**, typed OpenAI-compatible options and tool choice, JSON schema to grammar conversion"* — addressed #864 |
| [#1048](https://github.com/utilityai/llama-cpp-rs/pull/1048) | **closed, not merged** 2026-06 | *"unfortunately the latest releases **dropped all support for tool use**. this tries to add back some apis"* |

Verified against the 0.1.154 source actually on this machine:

```
grep -rl "tool" src/     → llguidance_sampler.rs, lib.rs, and a .gguf test fixture. That is all.
pub fn apply_chat_template(&self, tmpl, chat: &[LlamaChatMessage], add_ass: bool)
                                                  ↑ no tools parameter
pub fn json_schema_to_grammar(schema_json: &str)  ← this one survived
```

So **#977 was merged and then regressed away.** The only surviving piece is `json_schema_to_grammar` — which is, not coincidentally, the piece #485's PR 2 is built on.

That makes the ask *"please restore what you merged"* rather than *"please add this"*, with a merged PR to point at — a materially better position. It also means a restoration could regress again, which is an argument for #485's PRs 1–3 standing on their own rather than being written against a promise.

## Labelled `upstream`

Not actionable here. Nothing in Roteiro is blocked on it: #485's PRs 1–3 ship regardless, and PR 3 collapses into it only if it lands.

**Open question for a maintainer here:** whether to add the 153-byte measurement to upstream #1096. It is a real contribution to a report with no comments, but it posts publicly under this account, so I have not done it unasked.

**Comment by pixie79** [comment]
https://github.com/utilityai/llama-cpp-rs/pull/1048 the comments here are valid - this should be done outside of lamma and thus in this repo

**Comment by pixie79** [comment]
## Upstream has declined, with reasons — this is ours to build

Two maintainers on [utilityai/llama-cpp-rs#1048](https://github.com/utilityai/llama-cpp-rs/pull/1048), which was closed unmerged. The ask in this issue is not "not yet"; it is **out of scope by design**.

**@AsbjornOlling:**

> the goal of llama-cpp-rs is to be "dead simple, very close to raw bindings" … Handling chat templates, json parsing, tool call parsing and stuff like that is **in my opinion best done in rust, outside of llama.cpp**. It's all very high-level features, and I think this lib is best served by … operating on the level of logits and tokens, rather than on the level of messages.

They do exactly this in [NobodyWho](https://github.com/nobodywho-ooo/nobodywho): **minijinja** for chat templating, tool-call parsers implemented in Rust.

**@MarcusDunn:** *"I think tool calling can be handled in rust as @AsbjornOlling described much better than going through bindings to llama.cpp … my instinct is not to merge this."*

So the shim proposed above will not be accepted, and a fork is the only route through llama.cpp. Retitled and relabelled accordingly: this is Roteiro work, not an upstream wait.

## The part that is more than a scope argument

@AsbjornOlling gives concrete evidence that llama.cpp's implementations are **worse**, not merely out of scope — and three of the claims land directly on plans recorded in this repo:

| claim | where it lands here |
|---|---|
| llama.cpp chat templating is **slower than minijinja** ([benchmark](https://gitlab.com/AsbjornOlling/chat-template-benchmark)) | the premise of this issue was that upstream's engine is the thing worth reaching |
| **LLGuidance is significantly faster and more featureful** than llama.cpp's grammar sampler ([maskbench](https://github.com/guidance-ai/jsonschemabench/tree/main/maskbench)) | #485's design **rejected** `llguidance_sampler` on cost. That rejection was about per-request init, not mask throughput — the two may both be true, but it deserves re-testing rather than inheriting |
| **"String-based lazy grammars are extremely inefficient (they scale with the length of the full context)"** | #485 PR 2 plans `grammar_lazy_patterns` triggering on `<tool_call>` — i.e. exactly a string-based lazy grammar. Unmeasured here, and it compounds with #578 |
| GBNF parser bugs; *"a serious bug in upstream that causes a segfault with a significant subset of valid grammars"* | a reliability argument against the grammar route generally |
| llama.cpp context-shifts on a **token-count offset**, leaving messages split mid-sentence | not something this repo has looked at |

## What this changes

**#485 PR 3 no longer collapses into an upstream fix.** The dialect problem in #489 stays ours — which #515 already started, and the `Dialect::ALL` design there is the same architecture upstream is recommending: parse it in Rust.

**And the direction is already the one this project is taking.** #582/#583 are in flight doing precisely "handle the model's reasoning output in Rust rather than reaching for a C++ helper".

Research dispatched on the four open questions: whether minijinja can render our registry's real templates, what the full scope of doing this in Rust actually is, whether the lazy-grammar cost applies to us, and whether the LLGuidance rejection survives the maintainer's counter-evidence.

**Comment by pixie79** [comment]
## Q1 comes back clean — and it is not the blocker

**minijinja renders all five registry templates byte-identically to Python Jinja2.** 5 templates × 3 input shapes × tools/no-tools = 30 cells; pass/fail identical in all 30, and 16 of the 18 that render are byte-identical.

Oracle note, stated rather than glossed: llama.cpp's `common_chat_templates_apply` **could not** be compared — the binding does not exist, which is this issue's premise. Python **Jinja2 3.1.6 under `ImmutableSandboxedEnvironment`** was used instead, which is what `transformers` uses and what these templates were authored against. For *"does our real template render correctly"* that is the better oracle; llama.cpp's engine is itself a reimplementation chasing it.

**Zero unsupported constructs** — including the ones that usually break engines: `namespace()`, `loop.previtem`/`nextitem`, `macro`, `messages[::-1]` (Python slice with step), `|items`, `|tojson`, and Python string methods.

Three configuration decisions, each a one-liner, none a fork:

1. **`minijinja-contrib` `pycompat`** — without it `.startswith()` fails on `qwen3.8-27b:93`. The only reason bare minijinja fails at all.
2. **`preserve_order`** — without it object keys sort, so `{"query":…,"limit":…}` renders reordered. Silent, and it changes the prompt.
3. **Python-spaced `tojson`** — `{"a":1}` vs `{"a": 1}`; a 42-byte divergence on a two-tool surface. ~15 lines of `serde_json::ser::Formatter`.

One real semantic divergence: `none is iterable` is `True` in minijinja, `False` in Jinja2. It bites only if you pass `tools = null`. **Caller rule: omit, don't null** — which is what transformers does.

**3 net-new crates.** Render cost 3.6–24 µs, so the maintainer's speed benchmark is real and irrelevant here: against #578's 19-second prefill both engines are noise. Do not choose on that axis.

## The blocker is the `Message` type, not Jinja

`Message { role, content }` (`engine.rs:62`) has no `tool_calls`, no `tool_call_id`, no `name`. Every template renders assistant tool calls from **structured** fields; today the loop stuffs raw `<tool_call>…` markup into `content` as prose and lets ChatML pass it through.

So "render" is a **type change in `rto-llama`**, and that is the real scope of this issue.

## Three separable pieces, with the seams

| piece | state |
|---|---|
| **render** | the gap — blocked on `Message`, not on Jinja |
| **parse** | #515 done, but **2 of 3 dialects** — see #592 |
| **constrain** | does not exist. No grammar sampler anywhere in the tree; #485 closed but PR 2 never landed |
| **reasoning** | #582/#583 in flight |

## An ordering constraint that would otherwise be discovered late

`qwen3.8-27b`'s template ends its generation prompt with `<|im_start|>assistant\n<think>\n` — it **pre-opens the think block in the prompt**.

`strip_thinking` keys on `</think>`, so it survives. But `reviewer.rs:666` — `reasoning_truncated: reply.contains("<think>")` — **breaks silently**: under real rendering the opening tag lives in the prompt and never appears in the reply, so the field becomes permanently `false`.

That field exists specifically to catch the silent zero measured on `qwen3.8-27b`. **So #582/#583 must land before any render work**, or the render PR switches off the guard that #583 exists to provide, without failing anything.

## Two corrections to the claim table above

**The lazy-grammar warning is real but on the wrong axis, and it does *not* compound with #578.** Reading `llama_grammar_accept_impl:1394`, every accepted token's piece is appended to `trigger_buffer` and every trigger pattern is `regex_search`ed over the whole accumulation — O(N²), measured at 0.22 s / 2,048 tokens for a literal trigger and **0.75 s** for llama.cpp's optional-prefix form. But it scales with **generated** tokens, not context length: `trigger_buffer` starts empty per generation, and Roteiro never feeds prompt tokens to the sampler. **A 32 KiB tool surface adds 19 s of prefill and zero regex.** My table said these compound. They do not.

Pattern shape is a 3.4× lever if PR 2 is ever revived — use a literal.

**The LLGuidance rejection does not stand as stated, but is moot.** It priced per-request init the architecture never needed to pay — the binding's own doc says build once per model and reuse, and `rto-llama` already caches per-model derived objects in its residency entry (the mmproj, from #301). And "two new crates" was wrong by 4×: the real net-new count is **8**. Moot rather than overturned, because nothing in the tree uses a grammar at all.

## Is it worth building?

**Yes, but smaller than this issue describes, and not for the reason it gives.**

Against: `qwen3-32b`'s template instructs *exactly* Roteiro's prose convention — JSON inside `<tool_call>`. Real rendering buys close to nothing there, which is why 3 of 5 models already work.

For, on two models where the current path is **wrong** rather than suboptimal:

- `qwen3-coder-30b-a3b` and `qwen3.8-27b` templates instruct the **XML** form while Roteiro's prompt instructs JSON. The model is told two different things, and #515 cleans up after the collision. Rendering the real template makes instruction and training agree instead of parsing the disagreement.
- **`voxtral-mini-3b` is not ChatML**, so `apply_chat_template` substring-detects on a template with no `<|im_start|>` and produces a structurally wrong prompt before tools are considered.

**Smallest first PR:** render the model's real template for the *existing* `Message { role, content }` — no tools, no type change, no grammar. `rto-llama` only, keeping `resolve_chat_template_from`'s fallback ladder for models embedding none.
